recipes/openai.py

```diff
@@ -0,0 +1,354 @@
+    # Even if the model is None, this call is used as a healthcheck to verify access.
+    headers = {
+        "Authorization": f"Bearer {api_key}",
+        "OpenAI-Organization": org,
+    }
+    r = retry(
+        lambda: httpx.get(
+            "https://api.openai.com/v1/models",
+            headers=headers,
+        ),
+        n=1,
+        timeout_s=1,
+    )
+    if r.status_code == 422:
+        m = (
+            "Could not access api.openai.com -- 422 permission denied."
+            "Visit https://beta.openai.com/account/api-keys to check your API keys."
+        )
+        msg.fail(m)
+        sys.exit(-1)
+    elif r.status_code != 200:
+        m = "Error accessing api.openai.com" f"{r.status_code}: {r.text}"
+        msg.fail(m)
+        sys.exit(-1)
+
+    if model is not None:
+        response = r.json()["data"]
+        models = [response[i]["id"] for i in range(len(response))]
+        if model not in models:
+            e = f"The specified model '{model}' is not available. Choices are: {sorted(set(models))}"
+            msg.fail(e, exits=1)
+
+    return api_key, org
+
+
+def read_prompt_examples(
+    path: Optional[Path], *, example_class: PromptExample
+) -> List[PromptExample]:
+    if path is None:
+        return []
+    elif path.suffix in (".yml", ".yaml"):
+        return read_yaml_examples(path, example_class=example_class)
+    elif path.suffix == ".json":
+        data = srsly.read_json(path)
+        assert isinstance(data, list)
+        return [PromptExample(**eg) for eg in data]
+    else:
+        msg.fail(
+            "The --examples-path (-e) parameter expects a .yml, .yaml or .json file."
+        )
+        sys.exit(-1)
+
+
+def load_template(path: Path) -> jinja2.Template:
+    # I know jinja has a lot of complex file loading stuff,
+    # but we're not using the inheritance etc that makes
+    # that stuff worthwhile.
+    if not path.suffix == ".jinja2":
+        msg.fail(
+            "The --prompt-path (-p) parameter expects a .jinja2 file.",
+            exits=1,
+        )
+    with path.open("r", encoding="utf8") as file_:
+        text = file_.read()
+    return jinja2.Template(text, undefined=jinja2.DebugUndefined)
+
+
+def retry(
+    call_api: Callable[[], httpx.Response],
+    n: int,
+    timeout_s: int,
+    error_codes: List[int] = [429, 503],
+) -> httpx.Response:
+    """Retry a call to the OpenAI API if we get a non-ok status code.
+
+    This function automatically retries a request if it catches a response
+    with an error code in `error_codes`. The amount of timeout also increases
+    exponentially every time we retry.
+    """
+    assert n >= 0
+    assert timeout_s >= 1
+    r = call_api()
+    i = -1
+    # We don't want to retry on every non-ok status code. Some are about
+    # incorrect inputs, etc. and we want to terminate on those.
+    while i < n and r.status_code in error_codes:
```

Comment:
I noticed httpx.ReadTimeout errors as well. They seemed to break the flow just around 60 requests. 

```
Traceback (most recent call last):
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_exceptions.py", line 10, in map_exceptions
    yield
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/backends/sync.py", line 28, in read
    return self._sock.recv(max_bytes)
  File "/usr/lib/python3.8/ssl.py", line 1226, in recv
    return self.read(buflen)
  File "/usr/lib/python3.8/ssl.py", line 1101, in read
    return self._sslobj.read(len)
socket.timeout: The read operation timed out

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_transports/default.py", line 60, in map_httpcore_exceptions
    yield
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_transports/default.py", line 218, in handle_request
    resp = self._pool.handle_request(req)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_sync/connection_pool.py", line 253, in handle_request
    raise exc
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_sync/connection_pool.py", line 237, in handle_request
    response = connection.handle_request(request)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_sync/connection.py", line 90, in handle_request
    return self._connection.handle_request(request)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_sync/http11.py", line 112, in handle_request
    raise exc
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_sync/http11.py", line 91, in handle_request
    ) = self._receive_response_headers(**kwargs)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_sync/http11.py", line 155, in _receive_response_headers
    event = self._receive_event(timeout=timeout)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_sync/http11.py", line 191, in _receive_event
    data = self._network_stream.read(
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/backends/sync.py", line 28, in read
    return self._sock.recv(max_bytes)
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpcore/_exceptions.py", line 14, in map_exceptions
    raise to_exc(exc)
httpcore.ReadTimeout: The read operation timed out

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/usr/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/prodigy/__main__.py", line 61, in <module>
    controller = recipe(*args, use_plac=True)
  File "cython_src/prodigy/core.pyx", line 364, in prodigy.core.recipe.recipe_decorator.recipe_proxy
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/plac_core.py", line 367, in call
    cmd, result = parser.consume(arglist)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/plac_core.py", line 232, in consume
    return cmd, self.func(*(args + varargs + extraopts), **kwargs)
  File "recipes/openai_textcat.py", line 287, in textcat_openai_fetch
    srsly.write_jsonl(output_path, stream)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/srsly/_json_api.py", line 138, in write_jsonl
    for line in lines:
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/recipes/openai.py", line 129, in set_hashes
    for example in stream:
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/recipes/openai.py", line 175, in format_suggestions
    for example in stream:
  File "cython_src/prodigy/components/preprocess.pyx", line 164, in add_tokens
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/spacy/language.py", line 1545, in pipe
    for doc in docs:
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/spacy/language.py", line 1589, in pipe
    for doc in docs:
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/spacy/language.py", line 1586, in <genexpr>
    docs = (self._ensure_doc(text) for text in texts)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/spacy/language.py", line 1535, in <genexpr>
    docs_with_contexts = (
  File "cython_src/prodigy/components/preprocess.pyx", line 157, in genexpr
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/recipes/openai.py", line 159, in stream_suggestions
    responses = self._get_openai_response(prompts)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/recipes/openai.py", line 196, in _get_openai_response
    r = retry(
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/recipes/openai.py", line 320, in retry
    r = call_api()
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/recipes/openai.py", line 197, in <lambda>
    lambda: httpx.post(
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_api.py", line 304, in post
    return request(
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_api.py", line 100, in request
    return client.request(
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_client.py", line 821, in request
    return self.send(request, auth=auth, follow_redirects=follow_redirects)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_client.py", line 908, in send
    response = self._send_handling_auth(
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_client.py", line 936, in _send_handling_auth
    response = self._send_handling_redirects(
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_client.py", line 973, in _send_handling_redirects
    response = self._send_single_request(request)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_client.py", line 1009, in _send_single_request
    response = transport.handle_request(request)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_transports/default.py", line 218, in handle_request
    resp = self._pool.handle_request(req)
  File "/usr/lib/python3.8/contextlib.py", line 131, in __exit__
    self.gen.throw(type, value, traceback)
  File "/home/vincent/Development/prodigy-demos/openai-prodigy-recipes/venv/lib/python3.8/site-packages/httpx/_transports/default.py", line 77, in map_httpcore_exceptions
    raise mapped_exc(message) from exc
httpx.ReadTimeout: The read operation timed out
```